### PR TITLE
Automatically detect and track flaky tests

### PR DESCRIPTION
**Pull request by jasonrudolph**
**tl;dr** This pull request proposes configuring tooling to automatically detect and track flaky tests. As I understand it, this has historically required manual effort for Homebrew maintainers. By automating the process, we hopefully reduce the time that maintainers have to spend on the thankless and laborious task of manually identifying flaky tests. 🤞

/cc @MikeMcQuaid for review

## Pull request template

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
     _No. I don't think this is applicable to this pull request, but if it is, please let me know._ :bow:
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

## How it works

At a high-level, setting things up works as follows:

1. Enhance the test suite to emit JUnit XML test reports.

    This is accomplished in e163eb8650cb38d03829cd8aa537a3f32af3682f. Please see the commit message for details.

2. Enhance the CI workflow to send those test reports to [buildpulse.io](https://buildpulse.io) for analysis.

    This is accomplished in 006299dcc148bf1cdebddac55c7b1d581009b528.

3. Maintainers can access an always-up-to-date list of flaky tests, along with stats on the disruptiveness of each test, how recently it has occurred, etc. Maintainers can also choose to receive daily digests in Slack and weekly digests via email, so that they don't have to remember to check yet another tool. 😅

## A few things to have in mind

- 🐣 To date, BuildPulse has been used primarily by teams working in private repositories. Homebrew/brew would be the first significant open source project using BuildPulse, so **we'll probably discover some speedbumps** due to the differences in workflows between open source projects versus private projects. I hope you'll bring those speedbumps to my attention (jason@buildpulse.io), because I'd love to find a good way to address them.

- 🍴 For GitHub Actions, workflows currently need to use GitHub Actions Secrets in order to authenticate when sending test results to BuildPulse. Since forks can't access secrets, this means that **forks can't yet send test results to BuildPulse** for analysis. Other tools ([like CodeCov](https://github.com/Homebrew/brew/blob/aa7d3280d40074a6b927e6e9554db2964f1fa53d/.github/workflows/tests.yml#L230)) have a solution for this kind of thing, and I hope to add support for forks in BuildPulse in the future as well.

- 🍎 This pull request enables flaky test detection **exclusively for the [macOS CI job](https://github.com/Homebrew/brew/blob/006299dcc148bf1cdebddac55c7b1d581009b528/.github/workflows/tests.yml#L247-L250)**.

    CI also runs tests in three other contexts: ["tests (no-compatibility mode)", "tests (generic OS)", "tests (Linux)"](https://github.com/Homebrew/brew/blob/006299dcc148bf1cdebddac55c7b1d581009b528/.github/workflows/tests.yml#L187-L195). Before enabling BuildPulse for the other contexts, BuildPulse will need to evolve to accept additional metadata to let you distinguish which context reported a given test result.

    Right now, the XML reports produced by `rspec_junit_reporter` don't distinguish one context's results from another context's results. For reliably detecting flaky tests, if we're sending BuildPulse results for multiple contexts, it will be important to distinguish which context the test is running in.

    For example, let's assume that a given test (*x*) runs in both the "tests (Linux)" context and the "test everything (macOS)" context. And let's say we're running the build multiple times at commit `0abcdef`. If test *x* always passes in the macOS job, and always fails in the Linux job, the test isn't flaky; it's just broken on Linux. :grimacing:

    But if we don't know the context associated with each test result, all we'd see is that test *x* sometimes passes and sometimes fails at commit `0abcdef`, and we'd wrongly conclude that the test is flaky. :see_no_evil:

    So, for now, we need to pick one context to use for sending test results to BuildPulse. @MikeMcQuaid recommended using the macOS job, since it runs the largest number of tests (e.g. the `brew cask` ones).

## Verifying these changes

Since this pull request makes changes to the macOS CI job, and since that job doesn't run in forks, we can't immediately see the effect of these changes. To work around that limitation (so that reviewers can see the effect of these changes), 1888739cf69963323dfcc179440c6b0f10ff1c3b temporarily updates the macOS job to run in my fork. We can see that the job succeeds (https://github.com/jasonrudolph/brew/runs/2877954175) and that test results were successfully sent to BuildPulse. [[screenshot](https://user-images.githubusercontent.com/2988/122814647-7e463f80-d2a2-11eb-8138-19f2b2d9c4d0.png)]

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -320,3 +320,13 @@ jobs:
       - run: brew test-bot --only-formulae --test-default-formula
 
       - uses: codecov/codecov-action@29386c70ef20e286228c72b668a06fd0e8399192
+
+      - name: Upload test results to BuildPulse for flaky test detection
+        if: '!cancelled()' # Run this step even when the tests fail. Skip if the workflow is cancelled.
```
Comment (MikeMcQuaid) [inline .github/workflows/tests.yml:325]:
```suggestion
        # Only run this step for PRs where where we have access to secrets.
        # Run this step even when the tests fail. Skip if the workflow is cancelled.
        if: github.event.pull_request.head.repo.full_name == github.repository && !cancelled()
```

**Review by MikeMcQuaid — COMMENTED** [review]
These changes all seem good/reasonable to me and I'd love to see us use BuildPulse for this. I think we want a way to guard this to only PRs from Not Forks and `master` branch builds.

@Homebrew/brew any thoughts or objections on using BuildPulse for this?
@Homebrew/core I think this might be a good tool for us to eventually keep track of flaky `brew audit` and `brew test` runs for formulae, too.

**Comment by carlocab** [comment]
> These changes all seem good/reasonable to me and I'd love to see us use BuildPulse for this. I think we want a way to guard this to only PRs from Not Forks and `master` branch builds.

Agreed. I don't think we have the bandwidth to make use of fork data anyway, even if there were was information there that we can't get from CI runs here.

> @Homebrew/brew any thoughts or objections on using BuildPulse for this?

❤️

> @Homebrew/core I think this might be a good tool for us to eventually keep track of flaky `brew audit` and `brew test` runs for formulae, too.

❤️ for this too. We'll probably have to teach `test-bot` to produce test output in a format BuildPulse can consume.

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -320,3 +320,15 @@ jobs:
...
+
+      - name: Upload test results to BuildPulse for flaky test detection
+        # Only run this step for PRs where where we have access to secrets.
+        # Run this step even when the tests fail. Skip if the workflow is cancelled.
+        if: github.event.pull_request.head.repo.full_name == github.repository && !cancelled()
+        env:
+          BUILDPULSE_ACCESS_KEY_ID: ${{ secrets.BUILDPULSE_ACCESS_KEY_ID }}
+          BUILDPULSE_SECRET_ACCESS_KEY: ${{ secrets.BUILDPULSE_SECRET_ACCESS_KEY }}
+        run: |
+          curl -fsSL https://github.com/buildpulse/test-reporter/releases/latest/download/test-reporter-darwin-amd64 > ./buildpulse-test-reporter
```
Comment (SMillerDev) [inline .github/workflows/tests.yml:332]: Is this something we can package in homebrew-core instead?

Comment (carlocab) [reply .github/workflows/tests.yml:332]: Agreed. I can try to look into a formula, but I'm struggling to imagine what a decent formula test would be...

Comment (jasonrudolph) [reply .github/workflows/tests.yml:332]: > I can try to look into a formula, but I'm struggling to imagine what a decent formula test would be...

I'm new to homebrew formula tests, but would it make sense to run the downloaded binary with the `--help` flag and ensure that you get a zero exit status and/or verify that the usage text is emitted?

> The repo also won't pass our notability audit, but I think we can let this slide.

Can you point me toward that audit when you have a moment? Depending on how significant the necessary changes, I may be able to get something in place for this.

Comment (carlocab) [reply .github/workflows/tests.yml:332]: We don't like `--help` tests, hence my difficulty. We usually want something that tries to exercise some functionality. I guess I could mock a `junit` file (if I knew what those were), pass a fake `--account-id` and `--repository-id`, and see if we get the expected error. Would this be a terrible misuse of your API?

Our audit checks the number of watchers/stars/forks. I think it's one of at least 30, 30, or 75 respectively. I've starred your repo to help this along 😄 That said, I think the reason we want to `brew install` this is so that we use a binary that's versioned and checksummed. I'm sure I don't need to emphasise how important this is for code we're running in CI.

Comment (jasonrudolph) [reply .github/workflows/tests.yml:332]: > We usually want something that tries to exercise some functionality. I guess I could mock a `junit` file (if I knew what those were), pass a fake `--account-id` and `--repository-id`, and see if we get the expected error. Would this be a terrible misuse of your API?

My hunch is that this will probably work. 😅 Is it important that the test attempts to talk to the BuildPulse API? If not, then there are local errors that we could trigger (e.g., specifying an invalid directory as an arg) to verify the expected error message.

> I think the reason we want to brew install this is so that we use a binary that's versioned and checksummed.

Cool. That makes sense. Can we rely on the GitHub releases URL to identify the version (e.g., https://github.com/buildpulse/test-reporter/releases/download/v0.15.0/test-reporter-darwin-amd64)? 

I'm happy to start publishing checksums as release artifacts going forward if that would help.

Comment (carlocab) [reply .github/workflows/tests.yml:332]: > there are local errors that we could trigger

Yea, I think those will work. I don't imagine the formula tests would run too often (i.e. only about as often as you cut releases, maybe a bit more), but I'd still feel bad about abusing your API for this. (Unless you appreciate your API being tested for how it deals with invalid calls...)

>I'm happy to start publishing checksums as release artifacts going forward if that would help.

We still need a reliable way to verify the checksums, though, and without a Homebrew formula that would mean hardcoding it in the workflow and then occasionally updating it. That's pretty much the amount of work it'll take to maintain the formula, so we may as well make it available to our users while we're doing so.

Comment (jasonrudolph) [reply .github/workflows/tests.yml:332]: > > there are local errors that we could trigger
> 
> Yea, I think those will work.

Cool. Thanks for talking through this with me, @carlocab.

Probably the simplest error is to try to submit a non-existent directory:

```
$ test-reporter submit not-a-dir
<buildpulse> Current version: BuildPulse Test Reporter 0.15.0 (darwin be57499 go1.16.4)
<buildpulse> Initiating `submit`
<buildpulse> Received args: not-a-dir
<buildpulse> Using working directory: /usr/local/Homebrew

path is not a directory: not-a-dir

See more help with --help
```

Would that meet the need?

Comment (carlocab) [reply .github/workflows/tests.yml:332]: Yes. Homebrew/homebrew-core#79806

Thanks for talking me through this as well! By the way, I've built this so that version information appears as
```
❯ buildpulse-test-reporter --version
BuildPulse Test Reporter 0.15.0 (darwin Homebrew go1.16.5)
```
since it seemed like cloning your repo just to get the commit hash is a bit excessive. I can change it to actually reflect the proper commit hash, or some other string, if you prefer.

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -320,3 +320,15 @@ jobs:
...
+          chmod +x ./buildpulse-test-reporter
+          ./buildpulse-test-reporter submit Library/Homebrew/test/junit --account-id 1503512 --repository-id 53238813
```
Comment (MikeMcQuaid) [inline .github/workflows/tests.yml:334]: Now that https://github.com/Homebrew/homebrew-core/pull/79806 has been merged:

```suggestion
          brew install buildpulse-test-reporter
          buildpulse-test-reporter submit Library/Homebrew/test/junit --account-id 1503512 --repository-id 53238813
```

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks so much for your first contribution! Without people like you submitting PRs we couldn't run this project. You rock, @jasonrudolph!

**Comment by bayandin** [comment]
> > @Homebrew/core I think this might be a good tool for us to eventually keep track of flaky `brew audit` and `brew test` runs for formulae, too.
> 
> ❤️ for this too. We'll probably have to teach `test-bot` to produce test output in a format BuildPulse can consume.

That would be great! I guess we'll need to bring back junit output for homebrew-test-bot (was removed https://github.com/Homebrew/homebrew-test-bot/pull/355)

**Comment by MikeMcQuaid** [comment]
> I guess we'll need to bring back junit output for homebrew-test-bot (was removed [Homebrew/homebrew-test-bot#355](https://github.com/Homebrew/homebrew-test-bot/pull/355))

@bayandin Yes! It's relatively self-contained so shouldn't be too tricky. Is this something you might be able to pick up?

**Comment by bayandin** [comment]
> > I guess we'll need to bring back junit output for homebrew-test-bot (was removed [Homebrew/homebrew-test-bot#355](https://github.com/Homebrew/homebrew-test-bot/pull/355))
> 
> @bayandin Yes! It's relatively self-contained so shouldn't be too tricky. Is this something you might be able to pick up?

Will do!

**Comment by MikeMcQuaid** [comment]
@bayandin 😍 I'd consider limiting the output just to e.g. `audit` and `test` for now (or anything else that's flaky, maybe `install`?).

File: .github/workflows/tests.yml (hunk 1 above)
Comment (jasonrudolph) [reply .github/workflows/tests.yml]: It looks like this conditional prevented this step also from running on the `master` branch (https://github.com/Homebrew/brew/runs/2893711700?check_suite_focus=true), presumably because there was no pull request present for that build. I think we want this step to always run for builds on the master branch, so I'll open a pull request to adjust this conditional.

Comment (MikeMcQuaid) [reply .github/workflows/tests.yml]: That'd be great, thanks @jasonrudolph!

**Comment by bayandin** [comment]
> @bayandin 😍 I'd consider limiting the output just to e.g. `audit` and `test` for now (or anything else that's flaky, maybe `install`?).

Yep, I think I'll start with `audit` and `test` and then it should be pretty easy to adjust (at least I hope it will be so)

File: .github/workflows/tests.yml (hunk 1 above)
Comment (jasonrudolph) [reply .github/workflows/tests.yml]: Cool. I pushed up https://github.com/Homebrew/brew/pull/11586 with the proposed change.
